active/ItemsRepeater/ItemsRepeater-FirstVisibleIndex.md

```diff
@@ -0,0 +1,124 @@
+
+The [ListView](https://docs.microsoft.com/uwp/api/windows.ui.xaml.controls.listview?view=winrt-19041) and [GridView](https://docs.microsoft.com/uwp/api/windows.ui.xaml.controls.gridview?view=winrt-19041) controls already have properties to access the indices of their first and last visible items in the form of the [FirstVisibleIndex](https://docs.microsoft.com/uwp/api/windows.ui.xaml.controls.itemswrapgrid.firstvisibleindex?view=winrt-19041#Windows_UI_Xaml_Controls_ItemsWrapGrid_FirstVisibleIndex) and [LastVisibleIndex](https://docs.microsoft.com/uwp/api/windows.ui.xaml.controls.itemswrapgrid.lastvisibleindex?view=winrt-19041) properties on their respective base panels, [ItemsStackPanel](https://docs.microsoft.com/uwp/api/windows.ui.xaml.controls.itemsstackpanel?view=winrt-19041) and [ItemsWrapGrid](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.ItemsWrapGrid?redirectedfrom=MSDN&view=winrt-19041#Windows_UI_Xaml_Controls_ItemsWrapGrid_FirstVisibleIndex). 
+
+# Description
+<!-- Use this section to provide a brief description of the feature.
+For an example, see the introduction to the PasswordBox control 
+(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->
+
+FirstVisibleIndex, LastVisibleIndex,and RealizedItems are properties of ItemsRepeater that give you access to which indices of items within an ItemsRepeater are currently visible on the screen.
+
+The `FirstVisibleIndex` property returns the lowest index of the currently visible indices, which is the first visible index. The `LastVisibleIndex` property returns the highest index of the currently visible indices, which is the last visible index. For contiguous scenarios/layouts (uniform layouts), all items with indices between the first and last visible index are currently visible on the screen. For non-contiguous/non-uniform/nested layouts, the first and last visible index may not correlate to the first and last visible items on the screen.
+
+
+The `RealizedItems` property returns all items within the ItemsRepeater that are currently realized and visible on the screen, including items that may be cut off or not 100% visible. This property is helpful for determining which items are visible in non-contiguous and non-uniform layout scenarios.
+
+# Examples
+<!-- Use this section to explain the features of the API, showing
+example code with each description. The general format is: 
+  feature explanation,
+  example code
+  feature explanation,
+  example code
+  etc.-->
+  
+<!-- Code samples should be in C# and/or C++/WinRT -->
+
+<!-- As an example of this section, see the Examples section for the PasswordBox control 
+(https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples). -->
+
+```csharp
+// Check if an item is currently visible in a uniform layout.
```

Comment:
I think having a VisibleItems along with a RealizedItems property makes sense, and also brings attention to the fact that the two are different, which might be a common confusion. 

If it doesn't impact perf too much, I think that having FirstVisibleIndex and LastVisibleIndex properties _as well_ would be nice. There are use cases for the list of items and use cases for just the index values. If we don't have properties to get just the indices, it leaves a bit of work to figure out those values (if VisibleItems is not sorted). Thoughts?

Reply:
If these properties are useful outside of just enumerating all the visible items, then it sounds reasonable. We need to make sure we document that all items between first and last visible indices are not guaranteed to be visible (because we could have a non-contiguous layout). 

It would be good to understand the use cases where the first/last indices would be used without enumerating the items right after. 

ItemsStackPanel also has First/LastCacheIndex which are the first and last indices of RealizedItems.

Reply:
Added a more explicit note that items between first and last visible indices not guaranteed to be visible in non-contiguous layouts.

I think first and last visible indices are useful if someone wants to put special styles on something that appears at the top/bottom of the screen (like the item with the last visible index might have a "new" tag on it for messages or subscription feeds). 

I also think having the first and last visible indices is helpful for scaled animation scenarios, like the ItemsRepeater sample in the XCG that shows a bunch of colors and animates the items to get bigger as they go towards the middle. Knowing the first and last visible items would let you quickly calculate which item is going to be on the screen next (rather than searching for it in a list of realized items). Does that make sense?  

I think First/LastCacheIndex naturally fit in with these as well, and should be added. Thoughts?

Reply:
Since we do independent scrolling (scrolling does not happen on the UI thread), both those scenarios you mention will not work as expected if we used the indices. The indices will update when layout runs which would be too late since scrolling is happening independently. I think both those scenarios should be achievable using the EffectiveViewportChanged event instead, or using comp animations on the items themselves relative to the ScrollViewer offset they way the XCG sample is written if I remember correctly.

Reply:
I think an important point to consider is whether those are normal properties (without notifying when those change) or whether those would be dependency properties (with notifying about changes). 

It makes a huge difference if those values are DP or not, since in the non DP case, those would be "on demand". In the "on demand" case, accessing those values could work. However that of course would be less performant if used frequently, but more performant if unused.

Reply:
Yes, totally valid. I think that creating these properties as dependency properties would have too much of a negative perf impact. I'm leaning more towards just providing RealizedItems, VisibleItems, and First/Last Cache index. Would it make more sense to have CachedItems? I think you could easily find the cached items via the first/last cache index, but with the cache there may be more use cases to see explicitly where it starts and ends, rather than what's in it.  Open to other options.

Reply:
CachedItems concept in ItemsStackPanel is the same as RealizedItems discussed here. So we would just have VisibleItems and RealizedItems. I have not seen any scenarios so far where the indices are helpful outside of enumerating the items so perhaps we don't add the indices until there is a concrete scenario ? I think the harder part is knowing which are the visible and realized items which is solved by the VisibleItems, RealizedItems API.

Reply:
I agree with @ranjeshj, I think the difficult part is knowing the realized and more importantly the visible items.

Reply:
Ah okay, I think I misunderstood a comment above. I'm pushing a commit now that names RealizedItems and VisibleItems as the only two additions to ItemsRepeater.